spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/security/IapAuthenticationAutoConfiguration.java

```diff
@@ -0,0 +1,76 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.security;
+
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.security.oauth2.resource.servlet.OAuth2ResourceServerAutoConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoderJwkSupport;
+import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver;
+
+/**
+ * Autoconfiguration for extracting pre-authenticated user identity from Google Cloud IAP header.
+ *
+ * <p>The following conditions must be present for this configuration to take effect:
+ * <ul>
+ *   <li>{@code spring.cloud.gcp.security.iap.enabled} property is true. Note that a missing property is treated as
+ *   {@code false}. Because there is no custom code beyond what Spring Security OAuth provides, no dependency can be
+ *   used as a consistent signal for using IAP authentication, making an explicit property necessary.
+ *   <li>Spring Security OAuth Resource Server 5.0 or higher is present on the classpath.
+ * </ul>
+ * <p>If these conditions are met, a custom {@link BearerTokenResolver} and an ES256 web registry-based JWT token
+ * decoder beans are provided.
+ * <p>If a custom {@link WebSecurityConfigurerAdapter} is present, it must add {@code .oauth2ResourceServer().jwt()}
+ * customization to {@link org.springframework.security.config.annotation.web.builders.HttpSecurity} object. If no
+ * custom {@link WebSecurityConfigurerAdapter} is found,
+ * Spring Boot's default {@code OAuth2ResourceServerWebSecurityConfiguration} will add this customization.
+ *
+ * @author Elena Felder
+ * @since 1.1
+ */
+@Configuration
+@ConditionalOnProperty("spring.cloud.gcp.security.iap.enabled")
```

Comment:
Why no `matchIfMissing = true`  as we have in all other auto-configuration in this project?

This property also has to be added to the `additional-spring-configuration-metadata.json`, if you are not going to extract a fully blown `@ConfigurationProperties` class, as you mention with TODO below.

Reply:
I don't see that you have addressed this one.

Do you see my review at all?

Thanks

Reply:
So, this one is interesting. It's true that our normal practice is to have` matchIfMissing = true`, but our normal practice is to, at the same time, add @ConditionalOnClass that will invalidate the autoconfig whenever a required dependency is missing. IAP, however, no longer has any unique required dependencies -- the only dependency it needs is `spring-boot-starter-oauth2-resource-server`, which could be present on the classpath for unrelated reasons (for example, an application uses OAuth and also brings in `spring-cloud-gcp-autoconfigure` with a datastore starter).

I considered adding matchIfMissing and having the starter turn on the property in its application.properties, but that does not seem to be done anywhere else in the project, either. Do you think this is a good alternative?

Reply:
OK. Fine with me.

Since we really don't have a strict dependency and you are definitely right with some side-effect, so let it it as `false` by default!

Then we need explain in the Docs that this feature must be turned on explicitly.
On the other hand is there a choice when we really know that we are in GCP, so that is fully OK to have this IAP enabled by default? Are we really going to use some other JWT mechanism there ?

Thanks for explanation!

Reply:
I could see someone hosting an application on GCP but using Facebook or Twitter for authentication. IAP is more of an enterprise authN/authZ mechanism, since someone needs to set up users and groups for it to be useful.

I am writing documentation now; apparently that takes me even longer than writing code :\